community/primitive-collections/src/main/java/org/neo4j/concurrent/WorkSync.java

```diff
@@ -19,112 +19,302 @@
+                return unit.isCancelled();
+            }
+
+            @Override
+            public boolean isDone()
+            {
+                return unit.isDone();
+            }
+
+            @Override
+            public Object get() throws InterruptedException, ExecutionException
+            {
+                int tryCount = 0;
+                while ( !unit.isDone() )
+                {
+                    if ( Thread.interrupted() )
+                    {
+                        throw new InterruptedException();
+                    }
+                    tryCount++;
+                    checkFailure( tryDoWork( unit, tryCount, true ) );
+                }
+                unit.checkCancelled();
+                return null;
+            }
+
+            @Override
+            public Object get( long timeout, TimeUnit timeUnit )
+                    throws InterruptedException, ExecutionException, TimeoutException
             {
-                if ( lock.tryLock( tryCount < 10? 0 : 10, TimeUnit.MILLISECONDS ) )
+                Objects.requireNonNull( timeUnit );
+                if ( unit.isDone() )
+                {
+                    unit.checkCancelled();
+                    return null; // short-circuit potentially expensive calls to nanoTime.
+                }
+                long deadline = System.nanoTime() + timeUnit.toNanos( timeout );
+                int tryCount = 0;
+                while ( System.nanoTime() < deadline )
                 {
-                    try
+                    if ( Thread.interrupted() )
                     {
-                        doSynchronizedWork();
+                        throw new InterruptedException();
                     }
-                    finally
+                    if ( unit.isDone() )
                     {
-                        lock.unlock();
+                        unit.checkCancelled();
+                        return null;
                     }
+                    tryCount++;
+                    checkFailure( tryDoWork( unit, tryCount, true ) );
                 }
+                throw new TimeoutException();
+            }
+        };
+    }
+
+    private WorkUnit<Material,W> enqueueWork( W work )
+    {
+        WorkUnit<Material,W> unit = new WorkUnit<>( work, Thread.currentThread(), stackEnd );
+        unit.next = stack.getAndSet( unit ); // benign race, see reverse()
+        return unit;
+    }
+
+    private Throwable tryDoWork( WorkUnit<Material,W> unit, int tryCount, boolean otherwisePark )
+    {
+        if ( tryLock( tryCount, unit, otherwisePark ) )
+        {
+            WorkUnit<Material,W> batch = null;
+            try
+            {
+                batch = reverseAndFilterCancelled( stack.getAndSet( stackEnd ) );
+                return doSynchronizedWork( batch );
+            }
+            finally
+            {
+                unlock();
+                unparkAnyWaiters();
+                markAsDone( batch );
```

Comment:
it is a bit unfortunate that we on exception will mark some units as _done_ even though their work might not have been processed at all.